### PR TITLE
Add FXIOS-12510 [HNT - Search Bar] tapping on search bar experience

### PR DESCRIPTION
**Pull request by cyndichin**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-12510)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/27264)

## :bulb: Description
Add tapping on homepage search bar shows a scrim layer for zero search state. 

## :movie_camera: Demos
https://github.com/user-attachments/assets/4e9bcea1-315e-4296-b796-8d898dc5ff71

## :pencil: Checklist
- [x] I filled in the ticket numbers and a description of my work
- [x] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] I ensured unit tests pass and wrote tests for new code
- [ ] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [ ] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [ ] If needed, I updated documentation and added comments to complex code
- [ ] If needed, I added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by cyndichin** [comment]
@yoanarios seems like I get the unit test message even though I do have tests. Looks like it gets triggered for new files, it makes sense if we can remove that line of code and keep the ones with 0 test coverage?

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 1]
```diff
@@ -3493,6 +3522,35 @@ class BrowserViewController: UIViewController,
...
+        view.bringSubviewToFront(zeroSearchDimmingView)
+
+        NSLayoutConstraint.activate([
+            zeroSearchDimmingView.topAnchor.constraint(equalTo: contentContainer.topAnchor),
+            zeroSearchDimmingView.leadingAnchor.constraint(equalTo: contentContainer.leadingAnchor),
+            zeroSearchDimmingView.bottomAnchor.constraint(equalTo: contentContainer.bottomAnchor),
+            zeroSearchDimmingView.trailingAnchor.constraint(equalTo: contentContainer.trailingAnchor)
+        ])
+
+        UIView.animate(withDuration: 0.3) {
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:3539]: Kept constant here instead of UX struct since it's easier to make changes here when doing design review. Happy to move it to struct in future if that's preferred.

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 2]
```diff
@@ -1111,7 +1118,22 @@ class BrowserViewController: UIViewController,
         )
     }
 
+    /// If we are showing the homepage search bar, then we should hide the address toolbar
+    private func shouldHideToolbar() -> Bool {
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:1122]: @thatswinnie I didn't see a state for showing / hiding the toolbar, so I added this method and checks directly here in BVC. Let me know if there's a better place otherwise! We can do this in a follow up PR since I'll need to do a follow up PR to show the toolbar when we aren't in homepage.

**Comment by cyndichin** [comment]
Tried to keep this smaller, but it's a bit hard to test the UI end to end without the refactoring on the state and addition to middleware.

**Comment by yoanarios** [comment]
> @yoanarios seems like I get the unit test message even though I do have tests. Looks like it gets triggered for new files, it makes sense if we can remove that line of code and keep the ones with 0 test coverage?

I've opened a [PR](https://github.com/mozilla-mobile/firefox-ios/pull/27528) to remove the check if the file is not found, which was added for the edge case where the file was not added in XCTResults with no lines of code which shouldn't happened in the project

**Comment by cyndichin** [comment]
> > @yoanarios seems like I get the unit test message even though I do have tests. Looks like it gets triggered for new files, it makes sense if we can remove that line of code and keep the ones with 0 test coverage?
> 
> I've opened a [PR](https://github.com/mozilla-mobile/firefox-ios/pull/27528) to remove the check if the file is not found, which was added for the edge case where the file was not added in XCTResults with no lines of code which shouldn't happened in the project

thank you so much! 🔥

**Comment by cyndichin** [comment]
@mergify backport release/v141

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 3]
```diff
@@ -1111,7 +1118,22 @@ class BrowserViewController: UIViewController,
...
+            for: .homepage,
+            window: windowUUID
+        )?.searchState.shouldShowSearchBar ?? false
+        guard shouldShowSearchBar else { return false }
+        return true
+    }
+
     private func switchToolbarIfNeeded() {
+        guard !shouldHideToolbar() else {
+            addressToolbarContainer.isHidden = true
```
Comment (Foxbolts) [inline firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:1134]: I am not seeing where we unhide the address toolbar on orientation change (to landscape), I'm only seeing that we unhide it when showing the zero search view

**Review by Foxbolts — COMMENTED** [review]
Looks great @cyndichin, very nice!

Just one issue/question

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift (hunk 3 above)
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:1134]: thanks for reviewing!! This may change since I have a separate ticket for showing / hiding the address toolbar - https://mozilla-hub.atlassian.net/browse/FXIOS-12632

Any issue with keeping this ticket specific to tapping on search middle bar shows the new zero state and then can address this in the other ticket avoid this PR from getting to large?

I'm planning on pairing with Winnie to see if we can modify the toolbar state instead - https://github.com/mozilla-mobile/firefox-ios/pull/27505#discussion_r2161997452

Comment (Foxbolts) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:1134]: works for me 👍

**Comment by cyndichin** [comment]
@mergify rebase

**Comment by cyndichin** [comment]
Unable to rebase via mergify due to conflict with main in feature flag config with stories redesign. Rebase locally.
